### ISSUE TITLE
How to use lightning without any explicit train_dataloader?

### ISSUE THREAD

**Issue by MrinalJain17** [PyTorchLightning/pytorch-lightning#4499]
#### What is your question?

I'm working on Neural Style Transfer (the original paper by Gatys et. al.). The method is essentially specific to one style and one content image only, which can be loaded in the `__init__()` method itself. And there's no other data set required (training or validation). Once we have the style and content images, we iteratively generate the stylized image (and the logic resides in `training_step()`).

However, Lightning requires at minimum to define a `train_dataloader()`. How should I move forward now? Should I create a dummy dataloader? Is there a recommended way to get around this? 

#### What's your environment?

 - OS: Linux, Win
 - Packaging: conda
 - Version: 1.0.4

**Comment by Vichoko** [comment]
I think you can define a dummy `train_dataloader`, you need to be careful tho in `training_step `it will be iterated and the element will be passed to the `training_step` as the `batch` argument.

**Comment by tchaton** [comment]
Hey @MrinalJain17,

Thanks for raising such questions. Really interesting :)

My hacky way to resolve this: Create an infinite dataloader returning None with a length chosen based on your learning rate scheduling scheme, load both images directly into the model, perform optimisation in training_step.

I hope it helps.
T.C

**Comment by MrinalJain17** [comment]
> 
> 
> I think you can define a dummy `train_dataloader`, you need to be careful though in `training_step` it will be iterated and the element will be passed to the `training_step` as the `batch` argument.
> 
> If the training step involves only one step, maybe would be a good practice to pass the example through a list of one element in train_dataloader and retrieve it through the `batch` argument, instead of referencing a custom variable through the init and then calling it in the training_step.

Hi @Vichoko . Yes, what you recommended makes complete sense. Although the reason I want to load the images through init is because style transfer requires the computation of a gram matrix - which is an expensive operation (involves a forward pass through a VGG network). We do this for both the generated and style image at each iteraton.

However, since the style image remains the same throughout the process - its gram matrix can be pre-computed, which should effectively reduce the training time by half (since we now only pass the generated image through VGG to get its gram matrix, and use the pre-computed one to get the overall loss).



> 
> 
> Hey @MrinalJain17,
> 
> Thanks for raising such questions. Really interesting :)
> 
> My hacky way to resolve this: Create an infinite dataloader returning None with a length chosen based on your learning rate scheduling scheme, load both images directly into the model, perform optimisation in training_step.
> 
> I hope it helps.
> T.C

Hey @tchaton , thank a lot for your suggestion - I'm planning to do something very similar.

Something that's bothering me is that the purpose of such a dummy dataloader is literally so that the training loop in the lightning trainer works fine - it adds no other value. I was thinking maybe I can load the images (style and content) in a datamodule (instead of the LightningModule) and compute their gram matrix there (like in `setup()` maybe). Then, return that same gram matrix for every iteration (things like batch size and length would have to be set to 1, I guess). This way, atleast the datamodule will serve a purpose. It's still hacky of course 😅. I'll probably update how things go here on this thread.
